tapdb/sqlite.go

```diff
@@ -149,6 +157,87 @@ func NewSqliteStore(cfg *SqliteConfig) (*SqliteStore, error) {
 	return s, nil
 }
 
+// backupSqliteDatabase creates a backup of the given SQLite database.
+func backupSqliteDatabase(srcDB *sql.DB, dbFullFilePath string) error {
+	if srcDB == nil {
+		return fmt.Errorf("backup source database is nil")
+	}
+
+	// Create a database backup file full path from the given source
+	// database full file path.
+	//
+	// Get the current time and format it as a Unix timestamp in
+	// nanoseconds.
+	timestamp := time.Now().UnixNano()
+
+	// Add the timestamp to the backup name.
+	backupFullFilePath := fmt.Sprintf(
+		"%s.%d.backup", dbFullFilePath, timestamp,
+	)
+
+	log.Infof("Creating backup of database file: %v -> %v",
+		dbFullFilePath, backupFullFilePath)
+
+	// Backup the database
+	vacuumIntoQuery := "VACUUM INTO ?;"
+	stmt, err := srcDB.Prepare(vacuumIntoQuery)
+	if err != nil {
+		return err
+	}
+	defer stmt.Close()
+
+	_, err = stmt.Exec(backupFullFilePath)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// backupAndMigrate is a helper function that creates a database backup before
+// initiating the migration, and then migrates the database to the latest
+// version.
+func (s *SqliteStore) backupAndMigrate(mig *migrate.Migrate,
+	currentDbVersion fn.Option[uint],
+	maxMigrationVersion fn.Option[uint]) error {
+
+	// Determine if a database migration is necessary given the current
+	// database version and the maximum migration version known to the
+	// driver.
+	versionUpgradePending := false
+	currentDbVersion.WhenSome(func(currentV uint) {
+		maxMigrationVersion.WhenSome(func(maxV uint) {
+			versionUpgradePending = currentV < maxV
+		})
+	})
+
+	// If a database migration version upgrade is pending, create a backup
+	// of the database before starting the migration.
+	if versionUpgradePending && !s.cfg.SkipMigrationDbBackup {
+		log.Infof("Creating backup of database before applying " +
+			"migrations")
+		err := backupSqliteDatabase(s.DB, s.cfg.DatabaseFileName)
+		if err != nil {
+			return err
+		}
+	} else {
+		log.Infof("Skipping database backup before applying migrations")
+	}
+
+	// Migration to the target version.
+	err := mig.Up()
+	switch {
+	case errors.Is(err, migrate.ErrNoChange):
```

Comment:
The `ErrNoChange` error is already handled by `applyMigrations`, so we can just do `return mig.Up()`.